subprojects/core/src/main/java/org/gradle/api/services/internal/RegisteredBuildServiceProvider.java

```diff
@@ -157,8 +183,18 @@ public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
         return ExecutionTimeValue.changingValue(this);
     }
 
+    private ImmutableList<Consumer<? super RegisteredBuildServiceProvider<T, P>>> getStopActions() {
+        synchronized (this) {
```

Comment:
🤔 Since the provider reference is exposed to users, should it really synchronize on `this`?

Reply:
> should it really synchronize on `this`?

You're right, this might be an issue. I'm reluctant to add stuff to providers after that memory regression with evaluation context wrappers, but it should not be that bad this time.